### PR TITLE
[SM-186] Service Accounts - Overview

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This PR is for implementing the Service Accounts list overview for the Secrets Manager project.

## Code changes

- apps/web/src/locales/en/messages.json

New localization text

- bitwarden_license/bit-web/src/app/sm/layout/empty-list.component.html
bitwarden_license/bit-web/src/app/sm/layout/empty-list.component.ts
bitwarden_license/bit-web/src/app/sm/shared/sm-shared.module.ts

Making a new component for the empty list display/operations. 

- bitwarden_license/bit-web/src/app/sm/layout/new-menu.component.ts

Fixing to have new-menu implement angular OnDestory.

- bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html
bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.ts

Updating secrets list overview to use the new empty list component.

- bitwarden_license/bit-web/src/app/sm/service-accounts/responses/service-account.response.ts
bitwarden_license/bit-web/src/app/sm/service-accounts/service-account.service.ts
libs/common/src/models/view/service-account-view.ts

New service account service and the needed models/views.

- bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-list.component.html
bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-list.component.ts

Service account list overview implementation.

- bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-routing.module.ts
 bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts.component.html 
 bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts.component.ts 
 bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts.module.ts
  bitwarden_license/bit-web/src/app/sm/sm-routing.module.ts

All the necessary components and modules needed to get Service Accounts tab setup.

### Table Empty State

image

### Table View
 
image

### Item Drop-Downs

image

image

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html [hunk 1]
```diff
@@ -2,26 +2,15 @@
...
-</div>
+<ng-container *ngIf="secrets?.length == 0">
+  <sm-empty-list
+    [secretManagerType]="SecretManagerType.Secret"
+    emptyListTitle="secretsEmptyListTitle"
+    emptyListContent="secretsEmptyListMessage"
+    [showImportSecrets]="true"
+  >
+  </sm-empty-list>
+</ng-container>
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html:13]: This seems to be somewhat cumbersome to extend, and I think we should look into a more generic solution.

What do you think about something like this?

```html
<sm-empty-list *ngIf="secrets?.length == 0">
  <ng-container title>{{ "secretsEmptyListTitle" | i18n }}</ng-container>
  <ng-container description>{{ "secretsEmptyListMessage" | i18n }}</ng-container>
  <button bitButton buttonType="secondary">
    <i class="bwi bwi-plus" aria-hidden="true"></i>
    {{ "newSecret" | i18n }}
  </button>
  <button bitButton buttonType="secondary">
    <i class="bwi bwi-save tw-rotate-180" aria-hidden="true"></i>
    {{ "importSecrets" | i18n }}
  </button>
</sm-empty-list>
```

The template of the `EmptyListComponent` would be:

```html
<div
  class="tw-mx-auto tw-flex tw-flex-col tw-items-center tw-justify-center tw-pt-6 tw-text-center"
>
  <div class="tw-max-w-sm">
    <img class="no-items-image" aria-hidden="true" />
    <h3 class="tw-font-semibold">
      <ng-content select="[title]"></ng-content>
    </h3>
    <p>
      <ng-content select="[description]"></ng-content>
    </p>
  </div>
  <div class="tw-space-x-2">
    <ng-content select="[bitButton]"></ng-content>
  </div>
</div>
```

I'm also not sure if we need `List` in the name, it's more a `NoItemsComponent`.

File: bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-list.component.html [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        <span> 0 </span>
+      </td>
+      <td bitCell>{{ serviceAccount.revisionDate | date: "medium" }}</td>
+      <td bitCell>
+        <button
+          class="tw-rounded tw-border tw-border-transparent tw-bg-transparent tw-text-main hover:tw-border-secondary-700 focus:tw-border-secondary-700"
+          [bitMenuTriggerFor]="serviceAccountMenu"
+        >
+          <i class="bwi bwi-ellipsis-v tw-text-xl" aria-hidden="true"></i>
+        </button>
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-list.component.html:67]: This should use the icon button.

File: bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html (hunk 1 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html]: Looks great. 

Resolved in https://github.com/bitwarden/clients/pull/3653/commits/f435132f952de357418ffc1923b1d9c706b6566c

File: bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-list.component.html (hunk 2 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-list.component.html]: Resolved in https://github.com/bitwarden/clients/pull/3653/commits/f435132f952de357418ffc1923b1d9c706b6566c

**Comment by Thomas-Avery** [comment]
@Hinton  

It looks like we are importing both the `SecretsSharedModule` and the `SharedModule` in:

bitwarden_license/bit-web/src/app/sm/projects/projects.module.ts
bitwarden_license/bit-web/src/app/sm/secrets/secrets.module.ts
bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts.module.ts


Since `SecretsSharedModule` already imports `SharedModule`, should we just import the `SecretsSharedModule`? 

Like so https://github.com/bitwarden/clients/pull/3653/commits/d7b4c394a7b31413514f66d43e9f57ff3a1f878c

File: bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html [hunk 3]
```diff
@@ -2,37 +2,45 @@
...
-          [indeterminate]="selection.hasValue() && !isAllSelected()"
-        />
+        <div class="tw-flex tw-w-fit tw-gap-2">
+          <input
+            type="checkbox"
+            (change)="$event ? toggleAll() : null"
+            [checked]="selection.hasValue() && isAllSelected()"
+            [indeterminate]="selection.hasValue() && !isAllSelected()"
+          />
+          {{ "all" | i18n }}
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html:29]: Is this a design decision? It expands the column to be significant wider than before.

File: bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html [hunk 4]
```diff
@@ -2,37 +2,45 @@
...
         <button
-          class="tw-rounded tw-border tw-border-transparent tw-bg-transparent tw-text-main hover:tw-border-secondary-700 focus:tw-border-secondary-700"
+          bitIconButton="bwi-ellipsis-v"
+          buttonType="main"
+          title="{{ 'options' | i18n }}"
           [bitMenuTriggerFor]="tableMenu"
-        >
-          <i class="bwi bwi-ellipsis-v tw-text-xl" aria-hidden="true"></i>
-        </button>
+          [attr.title]="'options' | i18n"
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html:41]:
```suggestion
          [title]="'options' | i18n"
```

File: bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html [hunk 5]
```diff
@@ -56,11 +64,13 @@
...
         <button
-          class="tw-rounded tw-border tw-border-transparent tw-bg-transparent tw-text-main hover:tw-border-secondary-700 focus:tw-border-secondary-700"
+          bitIconButton="bwi-ellipsis-v"
+          buttonType="main"
           [bitMenuTriggerFor]="secretMenu"
-        >
-          <i class="bwi bwi-ellipsis-v tw-text-xl" aria-hidden="true"></i>
-        </button>
+          title="{{ 'options' | i18n }}"
+          [attr.title]="'options' | i18n"
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html:71]:
```suggestion
```

File: bitwarden_license/bit-web/src/app/sm/secrets/secrets.module.ts [hunk 6]
```diff
@@ -11,7 +9,7 @@ import { SecretsRoutingModule } from "./secrets-routing.module";
 import { SecretsComponent } from "./secrets.component";
 
 @NgModule({
-  imports: [SharedModule, SecretsRoutingModule, SecretsSharedModule],
+  imports: [SecretsRoutingModule, SecretsSharedModule],
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/secrets/secrets.module.ts:12]: Please put the routing module last.

File: bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-list.component.html [hunk 7]
```diff
@@ -31,11 +30,13 @@
       <th bitCell>{{ "lastEdited" | i18n }}</th>
       <th bitCell class="tw-w-0">
         <button
-          class="tw-rounded tw-border tw-border-transparent tw-bg-transparent tw-text-main hover:tw-border-secondary-700 focus:tw-border-secondary-700"
+          bitIconButton="bwi-ellipsis-v"
+          buttonType="main"
+          title="{{ 'options' | i18n }}"
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-list.component.html:35]:
```suggestion
```

File: bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-list.component.html [hunk 8]
```diff
@@ -31,11 +30,13 @@
...
           [bitMenuTriggerFor]="tableMenu"
-        >
-          <i class="bwi bwi-ellipsis-v tw-text-xl" aria-hidden="true"></i>
-        </button>
+          [attr.title]="'options' | i18n"
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-list.component.html:37]:
```suggestion
          [title]="'options' | i18n"
```

File: bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-list.component.html [hunk 9]
```diff
@@ -60,13 +61,14 @@
...
         <button
-          class="tw-rounded tw-border tw-border-transparent tw-bg-transparent tw-text-main hover:tw-border-secondary-700 focus:tw-border-secondary-700"
+          bitIconButton="bwi-ellipsis-v"
+          buttonType="main"
           [bitMenuTriggerFor]="serviceAccountMenu"
-        >
-          <i class="bwi bwi-ellipsis-v tw-text-xl" aria-hidden="true"></i>
-        </button>
+          title="{{ 'options' | i18n }}"
+          [attr.title]="'options' | i18n"
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-list.component.html:68]:
```suggestion
          title="{{ 'options' | i18n }}"
```

File: bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts.module.ts [hunk 10]
```diff
@@ -1,15 +1,13 @@
...
-
 import { SecretsSharedModule } from "../shared/sm-shared.module";
 
 import { ServiceAccountsListComponent } from "./service-accounts-list.component";
 import { ServiceAccountsRoutingModule } from "./service-accounts-routing.module";
 import { ServiceAccountsComponent } from "./service-accounts.component";
 
 @NgModule({
-  imports: [SharedModule, ServiceAccountsRoutingModule, SecretsSharedModule],
+  imports: [ServiceAccountsRoutingModule, SecretsSharedModule],
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts.module.ts:10]: Please put the routing module last.

File: bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html (hunk 3 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html:29]: Yes, all the FIGMAs have been updated to have "ALL" at the column header level.

File: bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts.module.ts (hunk 10 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts.module.ts]: Resolved in https://github.com/bitwarden/clients/pull/3653/commits/d12f9f9f9b62a60936ddffe1fc8feec5534baf7f

File: bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-list.component.html (hunk 8 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts-list.component.html]: Resolved in https://github.com/bitwarden/clients/pull/3653/commits/d12f9f9f9b62a60936ddffe1fc8feec5534baf7f

File: bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html (hunk 5 above)
Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/secrets/secrets-list.component.html]: Resolved in https://github.com/bitwarden/clients/pull/3653/commits/d12f9f9f9b62a60936ddffe1fc8feec5534baf7f

File: bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts.component.ts [hunk 11]
```diff
@@ -10,38 +10,26 @@ import { ServiceAccountService } from "./service-account.service";
   selector: "sm-service-accounts",
   templateUrl: "./service-accounts.component.html",
 })
-export class ServiceAccountsComponent implements OnInit, OnDestroy {
-  serviceAccounts: ServiceAccountView[];
+export class ServiceAccountsComponent implements OnInit {
+  serviceAccounts: Observable<ServiceAccountView[]>;
```
Comment (Hinton) [inline bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts.component.ts:14]: We should suffix this with `$`. Just noticed I forgot to change that in my previous PR 😢

Comment (Thomas-Avery) [reply bitwarden_license/bit-web/src/app/sm/service-accounts/service-accounts.component.ts]: Resolved in https://github.com/bitwarden/clients/pull/3653/commits/875ba3d6213c0b837670a3b19bcb76664fd5e709

**Review by Hinton — APPROVED** [review]

**Pull request merged**